File: ios/Classes/proxy/RtpSenderProxy.swift
import WebRTC

/// Wrapper around an `RTCRtpSender`, powering it with additional API.
class RtpSenderProxy {
  /// Actual underlying `RTCRtpSender`.
  private var sender: RTCRtpSender

  /// `MediaStreamTrackProxy` of this `RtpSenderProxy`.
  private var track: MediaStreamTrackProxy?

  /// Initializes a new `RtpSenderProxy` for the provided `RTCRtpSender`.
  init(sender: RTCRtpSender) {
    self.sender = sender
    self.syncMediaStreamTrack()
  }

  /// Returns ID of this `RtpSenderProxy`.
  func id() -> String {
    self.sender.senderId
  }

  /// Replaces the `MediaStreamTrackProxy` of the underlying `RtpSender` with
  /// the provided one.
  func replaceTrack(t: MediaStreamTrackProxy?) {
    self.track = t
    self.sender.track = t?.obj()
  }

  /// Returns `RtpParameters` of the underlying `RtpSender`.
  func getParameters() -> RTCRtpParameters {
    return self.sender.parameters
  }

  /// Sets `RtpParameters` of the underlying `RtpSender` with the provided one.
  func setParameters(params: RTCRtpParameters) {
    self.sender.parameters = params
  }

  /// Synchronizes the `MediaStreamTrackProxy` of this `RtpSenderProxy` with the
  /// underlying `RtpSender`.
  func syncMediaStreamTrack() {
    let newTrack = self.sender.track
    if newTrack == nil {
      self.track = nil
    } else {
      if self.track == nil {
        self.track = MediaStreamTrackProxy(
          track: newTrack!,
          deviceId: nil,
          source: nil
        )
      }
    }
  }
}

File: android/src/main/kotlin/com/instrumentisto/medea_flutter_webrtc/proxy/RtpSenderProxy.kt
package com.instrumentisto.medea_flutter_webrtc.proxy

import com.instrumentisto.medea_flutter_webrtc.exception.ReplaceSenderTrackException
import org.webrtc.RtpParameters
import org.webrtc.RtpSender

/**
 * Wrapper around an [RtpSender].
 *
 * @param sender Actual underlying [RtpSender].
 */
class RtpSenderProxy(sender: RtpSender) : Proxy<RtpSender>(sender) {

  /** [MediaStreamTrackProxy] of this [RtpReceiverProxy]. */
  private var track: MediaStreamTrackProxy? = null

  /** Disposed state of the [obj]. */
  private var disposed: Boolean = false

  init {
    syncMediaStreamTrack()
    addOnSyncListener { syncMediaStreamTrack() }
  }

  /** Sets [disposed] to `true` for the [obj]. */
  fun setDisposed() {
    disposed = true
  }

  /**
   * Replaces [MediaStreamTrackProxy] of the underlying [RtpSender] with the provided one.
   *
   * @param t [MediaStreamTrackProxy] which will be set to the underlying [RtpSender].
   */
  fun replaceTrack(t: MediaStreamTrackProxy?) {
    if (disposed) {
      return
    }

    track = t
    val isSuccessful = obj.setTrack(t?.obj, false)
    if (!isSuccessful) {
      throw ReplaceSenderTrackException()
    }
  }

  /** Returns [RtpParameters] of the underlying [RtpSender]. */
  fun getParameters(): RtpParameters {
    return obj.parameters
  }

  /**
   * Sets [RtpParameters] of the underlying [RtpSender] with the provided one.
   *
   * @return [Boolean] indicating whether parameters has been set.
   */
  fun setParameters(params: RtpParameters): Boolean {
    return obj.setParameters(params)
  }

  /**
   * Synchronizes the [MediaStreamTrackProxy] of this [RtpSenderProxy] with the underlying
   * [RtpSender].
   */
  private fun syncMediaStreamTrack() {
    val newSenderTrack = obj.track()
    if (newSenderTrack == null) {
      track = null
    } else {
      if (track == null) {
        track = MediaStreamTrackProxy(newSenderTrack)
      } else {
        track!!.replace(newSenderTrack)
      }
    }
  }
}
